relstorage/cache/ring.c

```diff
@@ -0,0 +1,435 @@
+}
+
+void lru_on_hit(RSRing ring, RSRingNode* entry)
+{
+    entry->frequency++;
+    ring_move_to_head(ring, entry);
+}
+
+static int lru_will_fit(RSRingNode* ring, RSRingNode* entry)
+{
+    return ring->max_len >= (entry->len + ring->frequency);
+}
+
+/**
+ * Call when `updated_ring` has gotten too big, and we should transfer
+ * items to `destination_ring`. The item `ignore_me` is the item that
+ * caused the `updated_ring` to get too big, so if it's the only thing
+ * in the ring, we don't move it.
+ *
+ * When `allow_victims` is False, then we stop once we fill up all
+ * three rings and we avoid producing any victims. If we *would*
+ * have produced victims, we return with rejects.frequency = 1 so the
+ * caller can know to stop feeding us.
+ *
+ * When `overfill_destination` is True, then we will move all the
+ * items we have to in order to make updated_ring its correct size,
+ * even though this could cause the destination to become too large.
+ * There will never be victims in this case. `allow_victims` is
+ * ignored in this case.
+ *
+ * You must only call this when the ring is already full.
+ */
+#define _SPILL_OVERFILL 1
+#define _SPILL_FIT 0
+#define _SPILL_VICTIMS 1
+#define _SPILL_NO_VICTIMS 0
+static
+RSRingNode _spill_from_ring_to_ring(RSRing updated_ring,
+                                    RSRing destination_ring,
+                                    RSRingNode* ignore_me,
+                                    int allow_victims,
+                                    int overfill_destination)
+{
+    RSRingNode rejects = {};
+    if(overfill_destination) {
+        rejects.r_next = rejects.r_prev = NULL;
+    }
+    else {
+        rejects.r_next = rejects.r_prev = &rejects;
+    }
+
+    while(updated_ring->len > 1 && ring_oversize(updated_ring)) {
+        RSRingNode* eden_oldest = ring_lru(updated_ring);
+        if(!eden_oldest || eden_oldest == ignore_me) {
+            break;
+        }
+
+        if(overfill_destination || lru_will_fit(destination_ring, eden_oldest)){
+            // Good, there's room. No victims to choose.
+            ring_move_to_head_from_foreign(updated_ring, destination_ring, eden_oldest);
+        }
+        else {
+            // Darn, we're too big. We must choose (and record) a
+            // victim.
+
+            if(!allow_victims) {
+                // set the signal and quit.
+                rejects.frequency = 1;
+                break;
+            }
+
+            RSRingNode* probation_oldest = ring_lru(destination_ring);
+            if(!probation_oldest) {
+                //Hmm, the ring got emptied, but there's also no space
+                //in protected. This must be a very large object. Take
+                //ownership of it anyway, but quite trying.
+               ring_move_to_head_from_foreign(updated_ring, destination_ring, eden_oldest);
+               break;
+            }
+
+            if(probation_oldest->frequency > eden_oldest->frequency) {
```

Comment:
Can you try `>=` as well? In my simulations that sometimes had a much better hit rate for workloads, especially MRU skewed ones (e.g. scans). When I tweak the simulator's TinyLfu condition and run at 512,

|condition|multi1|gli|
|---|---|---|
|candidateFreq >= victimFreq | 50.50% | 16.11%
|candidateFreq > victimFreq | 55.59% | 34.16%

Tthe victim is aged out due to a low frequency, so `c >= v` results in more admissions. This flushes the probation space faster, leaving less time for a probation entry to be reused and promoted to protected. The penalty was very tiny for the traces negatively impacted.

I originally wrote it like you did and then tried to optimize when evaluating policies. I was really surprised when I discovered that small change made such a difference. It appears to let the cache handle MRU (scans) much better.